src/utils/Assets.ts

```diff
@@ -0,0 +1,350 @@
+      if (tokenTemplate.isActive && id.toString() === template.toString()) {
+        return i
+      }
+    } else if (
+      tokenTemplate.isActive &&
+      tokenTemplate.templateAddress === template.toString()
+    ) {
+      return i
+    }
+  }
+  // if nothing is found it returns -1
+  return -1
+}
+/**
+ *
+ * @param name asset name
+ * @param symbol asse symbol
+ * @param owner owner address
+ * @param assetUrl asset url, if present and confidential evm, add it to token create params
+ * @param templateIDorAddress either template address or id
+ * @param ddo ddo
+ * @param encryptDDO encrypt or not?
+ * @param providerUrl the provider URL
+ * @param providerFeeToken the provider fee token
+ * @param aquariusInstance aquarius, could be node instance url
+ * @param allowAccessList?: string,
+ * @param denyAccessList?: string
+ * @returns ddo id as string
+ */
+export async function createAsset(
+  name: string,
+  symbol: string,
+  owner: Signer,
+  assetUrl: any, // files object
+  templateIDorAddress: string | number, // If string, it's template address , otherwise, it's templateId
+  ddo: any,
+  encryptDDO: boolean = true, // default is true
+  providerUrl: string,
+  providerFeeToken: string,
+  aquariusInstance: Aquarius,
+  accessListFactory?: string, // access list factory address
+  allowAccessList?: string, // allow list address
+  denyAccessList?: string // deny list address
+): Promise<string> {
+  const isAddress = typeof templateIDorAddress === 'string'
+  const isTemplateIndex = typeof templateIDorAddress === 'number'
+  if (!isAddress && !isTemplateIndex) {
+    throw new Error('Invalid template! Must be a "number" or a "string"')
+  }
+  const chainID = (await owner.provider.getNetwork()).chainId
+
+  const config = new ConfigHelper().getConfig(parseInt(String(chainID)))
+
+  let templateIndex = await calculateActiveTemplateIndex(
+    owner,
+    config.nftFactoryAddress,
+    templateIDorAddress
+  )
+
+  if (templateIndex < 1) {
+    // for testing purposes only
+    if (chainID === DEVELOPMENT_CHAIN_ID) {
+      templateIndex = 1
+    } else throw new Error(`Invalid template index: ${templateIndex}`)
+  }
+
+  const nft = new Nft(owner, chainID)
+
+  const nftFactory = new NftFactory(config.nftFactoryAddress, owner)
+
+  // get nft owner
+  const account = await owner.getAddress()
+
+  // from hex to number format
+  ddo.chainId = parseInt(chainID.toString(10))
+  const nftParamsAsset: NftCreateData = {
+    name,
+    symbol,
+    templateIndex: 1,
+    tokenURI: 'aaa',
+    transferable: true,
+    owner: account
+  }
+  const datatokenParams: DatatokenCreateParams = {
+    templateIndex,
+    cap: '100000',
+    feeAmount: '0',
+    paymentCollector: account,
+    feeToken: providerFeeToken || config.oceanTokenAddress,
+    minter: account,
+    mpFeeAddress: ZERO_ADDRESS
+  }
+
+  // include fileObject in the DT constructor
+  if (config.confidentialEVM) {
```

Comment:
**Observation**: If datatoken 4 has the file object and we want to create a bundled tx (meaning deploy NFT, datatoken 4 and pricing schema at once) and files object needs to be encrypted on chain, then we should also have a test when we check, if confidentialEVM flag is set on true we need to `sapphire.wrap()` the owner wallet address, otherwise the tx won't be encrypted on sapphire chain.
**Question**: When creating a datatoken from a standalone NFT, NFT class should also have this confidentialEVM flag when creating a datatoken template 4, right?

Reply:
for question, I have created https://github.com/oceanprotocol/ocean.js/pull/1859 in case.

Reply:
`sapphire.wrap()` is a specific API for Oasis. At this point its true that its the only confidential EVM we have, but that could be different tomorrow, we **cannot**  do shapire.wrap() on a TX just because its "confidentialEVM". It only works for Oasis shappire. In any case i think that applies only for `calldata` when interacting with the SC, as the smart contract should run on their own secure enclave... not really sure if that's on the scope of this function here. Maybe @alexcos20  can have some input here.. its easy, but in any case we need to check the chain, its not enough see if its confidential or not.

Regarding your question, i'm not sure if i understand your point.. When we have a confidential EVM, the requirement would be pass the filesObject on the DatatokenCreateParams

Reply:
> for question, I have created #1859 in case.

i see, but you're again doing a fixed match:  confidentialEVM === Shappire 
afaik, that might not always be the case

Reply:
I agree @paulo-ocean to have a stronger condition and include chain id as well. I wanted to raise an awareness that when creating a datatoken (either bundled or standalone) to encrypt the tx since we pass the files object, even though files object is already encoded in utf8 bytes, that was related to my question. I agree to also merge your PR and do this `sapphire.wrap()` thing separate in my PR if needed when creating a datatoken.

Reply:
updated my PR with the stronger condition and tests as well.